### PR TITLE
port toranj test21_address_cache_table

### PR DESCRIPTION
**Pull request by priyankaCh22**

### CODE REVIEW

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+    def test04_verify_address_cache_table_on_r1(self):
+        # Send a single UDP message from r1 to sed2
+        # Send a single UDP message from r1 to fed3
+        self.transmit_receive_udp_message([(self.r1, self.sed2, self.r1_address, self.sed2_address),
+                                           (self.r1, self.fed3, self.r1_address, self.fed3_address)])
+        time.sleep(WAIT_TIME)
+        # The address cache table on r1 should contain two entries for sed2 and fed3 addresses.
+
+        addr_cache_table = self.r1.wpanctl("get", "get " + wpan.WPAN_THREAD_ADDRESS_CACHE_TABLE, 2)
+        addr_cache_table = wpan_table_parser.parse_address_cache_table_result(addr_cache_table)
```
Comment (priyankaCh22) [inline silk/tests/openthread/ot_test_address_cache_table.py:210]: here parse_addess_cache_table is failing as out of 2 entries one entry is getting age:1 however the age:0 is required.

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+import random
+import time
+import unittest
+
+from silk.config import wpan_constants as wpan
+from silk.node.wpan_node import WpanCredentials
+from silk.utils import process_cleanup
+from silk.tools import wpan_table_parser
+from silk.tools.wpan_util import verify, verify_within, VerifyError
+from silk.unit_tests.test_utils import random_string
```
Comment (tzm41) [inline silk/tests/openthread/ot_test_address_cache_table.py]: Should sort import statements.

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        for device in cls.device_list:
+            device.tear_down()
+
+    @testcase.setup_decorator
+    def setUp(self):
+        pass
+
+    @testcase.teardown_decorator
+    def tearDown(self):
+        pass
```
Comment (tzm41) [inline silk/tests/openthread/ot_test_address_cache_table.py:102]: Are these needed if they don't do anything?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+        TestAddressCacheTable.fed3_address = self.fed3.find_ip6_address_with_prefix(PREFIX)
+
+    def transmit_receive_udp_message(self, addresses: list):
+        timeout = 5
+        delay = 1
+
+        for i, (src, dst, src_address, dst_address) in enumerate(addresses):
+            port = random.randint(10000 + i * 100, 10099 + i * 100)
+            message = random_string(10)
+            # src_address = f"{src_addr}%{src.netns}"
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: Please delete the commented out line

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+        self.r2.allowlist_node(self.r1)
+
+        self.r2.allowlist_node(self.sed2)
+        self.sed2.allowlist_node(self.r2)
+
+        self.r2.allowlist_node(self.r3)
+        self.r3.allowlist_node(self.r2)
+
+        self.r3.allowlist_node(self.fed3)
+        self.fed3.allowlist_node(self.r3)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: Put these lines into a loop ?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+        self.assertTrue(fed3_parent == r3_ext_address)
+
+    @testcase.test_method_decorator
+    def test03_verify_route_from_r1_to_r3(self):
+        # Wait till we have a valid "next hop" route on r1 towards r3
+        self.r1.add_prefix(PREFIX, stable=True, on_mesh=True, slaac=True, preferred=True)
+        self.wait_for_completion(self.device_list)
+
+        # Wait for prefix to get added to all the nodes
+        time.sleep(60)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: put it in a constant variable PREFIX_PROPAGATION_DELAY ?

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 1 above)
Comment (qilinkang) [reply silk/tests/openthread/ot_test_address_cache_table.py:210]: Have you figured out what went wrong ?

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 2 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 3 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py:102]: Will check and update on this.

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 4 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 5 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 6 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 1 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py:210]: Not yet. The pairing and router table are proper and it shows all the links from r1 to r3. And it is not even related to delay testing by adding 180sec delay and that too did not affect the behavior. Digging more into it.

Comment (qilinkang) [reply silk/tests/openthread/ot_test_address_cache_table.py:210]: Making changes for the following seems getting test_04 Pass.

class AddressCacheEntry(object):
     ....
      self._age = int(items_dict["Age"], 0)

def parse_address_cache_table_result(addr_cache_table_list):
    """Parses address cache table list string and returns an array of `AddressCacheEntry` objects.
    """
    return [AddressCacheEntry(item) for item in addr_cache_table_list.split("\n")[1:-2]]

Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py:210]: Thanks Qilin. Yes it was indeed a parsing issue.

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 3 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py:102]: yes these are required as without these test cases do not start. i.e. after setup class directly teardown class is executed.

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 7]
```diff
@@ -233,203 +220,208 @@ def test04_verify_address_cache_table_on_r1(self):
...
+            child_table = self.r2.wpanctl("get", "get " + wpan.WPAN_THREAD_CHILD_TABLE, 2)
+            child_table = wpan_table_parser.parse_child_table_address_result(child_table)
+            verify(len(child_table) == 0)
+
+        verify_within(check_sed2_is_removed_from_r2_child_table, REATTACH_WAIT_TIME)
+
+        # Verify that both sed2, fed3 are children of r3
+        child_table = self.r3.wpanctl("get", "get " + wpan.WPAN_THREAD_CHILD_TABLE, 2)
+        child_table = wpan_table_parser.parse_child_table_address_result(child_table)
+        self.assertEqual(len(child_table), 2)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: use verify() ?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 8]
```diff
@@ -233,203 +220,208 @@ def test04_verify_address_cache_table_on_r1(self):
...
+        # receives the IPv6 message with sed2's address, r2 itself would do an address
+        # query for the address and forward the IPv6 message.
+
+        self.transmit_receive_udp_message([(self.r1, self.sed2, self.r1_address, self.sed2_address)])
+        time.sleep(ADDRESS_CACHE_TABLE_WAIT_TIME)
+
+        # The address cache table on r1 should have sed2's address removed.
+        addr_cache_table = self.r1.wpanctl("get", "get " + wpan.WPAN_THREAD_ADDRESS_CACHE_TABLE, 2)
+        addr_cache_table = wpan_table_parser.parse_address_cache_table_result(addr_cache_table)
+        self.assertEqual(len(addr_cache_table), 1)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: use verify() ?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 9]
```diff
@@ -233,203 +220,208 @@ def test04_verify_address_cache_table_on_r1(self):
...
+    def test07_verify_address_cache_table_on_r1_adds_sed2(self):
+        # Send a UDP message from r1 to sed2.
+        self.transmit_receive_udp_message([(self.r1, self.sed2, self.r1_address, self.sed2_address)])
+
+        time.sleep(ADDRESS_CACHE_TABLE_WAIT_TIME)
+        # The address cache table on r1 should have both fed1 and sed2.
+
+        addr_cache_table = self.r1.wpanctl("get", "get " + wpan.WPAN_THREAD_ADDRESS_CACHE_TABLE, 2)
+        addr_cache_table = wpan_table_parser.parse_address_cache_table_result(addr_cache_table)
+        self.assertEqual(len(addr_cache_table), 2)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: use verify() ?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 10]
```diff
@@ -233,203 +220,208 @@ def test04_verify_address_cache_table_on_r1(self):
...
+            child_table = self.r3.wpanctl("get", "get " + wpan.WPAN_THREAD_CHILD_TABLE, 2)
+            child_table = wpan_table_parser.parse_child_table_address_result(child_table)
+            verify(len(child_table) == 1)
+
+        verify_within(check_sed2_is_removed_from_r3_child_table, REATTACH_WAIT_TIME)
+
+        # Verify that sed2 is a child of r2
+        child_table = self.r2.wpanctl("get", "get " + wpan.WPAN_THREAD_CHILD_TABLE, 2)
+        child_table = wpan_table_parser.parse_child_table_address_result(child_table)
+        self.assertEqual(len(child_table), 1)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: use verify() ?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 11]
```diff
@@ -233,203 +220,208 @@ def test04_verify_address_cache_table_on_r1(self):
...
+        self.transmit_receive_udp_message([(self.sed2, self.fed1, self.sed2_address, self.fed1_address)])
+        time.sleep(ADDRESS_CACHE_TABLE_WAIT_TIME)
+
+        # r1 upon receiving and forwarding the message from sed2 (through r2 now) should
+        # update its address cache table for sed2 (address cache update through snooping).
+        #
+        # verify that the address cache table is updated correctly.
+        addr_cache_table = self.r1.wpanctl("get", "get " + wpan.WPAN_THREAD_ADDRESS_CACHE_TABLE, 2)
+        addr_cache_table = wpan_table_parser.parse_address_cache_table_result(addr_cache_table)
+        self.assertEqual(len(addr_cache_table), 2)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: use verify() ?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 12]
```diff
@@ -233,203 +220,208 @@ def test04_verify_address_cache_table_on_r1(self):
...
+            child_table = self.r2.wpanctl("get", "get " + wpan.WPAN_THREAD_CHILD_TABLE, 2)
+            child_table = wpan_table_parser.parse_child_table_address_result(child_table)
+            verify(len(child_table) == 0)
+
+        verify_within(check_sed2_is_removed_from_r2_child_table, REATTACH_WAIT_TIME)
+
+        # Verify that both sed2, fed3 are children of r3
+        child_table = self.r3.wpanctl("get", "get " + wpan.WPAN_THREAD_CHILD_TABLE, 2)
+        child_table = wpan_table_parser.parse_child_table_address_result(child_table)
+        self.assertEqual(len(child_table), 2)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: use verify() ?

File: silk/tests/openthread/ot_test_address_cache_table.py [hunk 13]
```diff
@@ -233,203 +220,208 @@ def test04_verify_address_cache_table_on_r1(self):
...
+        # update its address cache table for sed2 (address cache update through snooping).
+        #
+        # verify that the address cache table is updated correctly.
+        # The address cache table on r1 should have sed2's address removed.
+
+        # The address cache table on r1 should have both fed1 and sed2.
+
+        addr_cache_table = self.r1.wpanctl("get", "get " + wpan.WPAN_THREAD_ADDRESS_CACHE_TABLE, 2)
+        addr_cache_table = wpan_table_parser.parse_address_cache_table_result(addr_cache_table)
+        self.assertEqual(len(addr_cache_table), 2)
```
Comment (qilinkang) [inline silk/tests/openthread/ot_test_address_cache_table.py]: use verify() ?

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 7 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 8 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 9 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 10 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 12 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

File: silk/tests/openthread/ot_test_address_cache_table.py (hunk 13 above)
Comment (priyankaCh22) [reply silk/tests/openthread/ot_test_address_cache_table.py]: Ack

**Review by qilinkang — APPROVED** [review]
LGTM

**Review by tzm41 — APPROVED** [review]
LGTM

**Review by jwhui — APPROVED** [review]

**Pull request merged**